### ISSUE TITLE
ed25519: Sub-optimal serialization for type-tagged formats

### ISSUE THREAD

**Issue by connec** [RustCrypto/signatures#336]
As of #78, the encoding for formats that "serialize byte slices more compactly or more efficiently than other types of slices" (to quote [serde's docs](https://docs.rs/serde/1.0.126/serde/trait.Deserializer.html#tymethod.deserialize_bytes)) may be worse.

For example, using Message Pack (via [`rmp_serde`](https://docs.rs/rmp-serde/)), the tuple-based approach leads to encoding as an array where each element is individually encoded, hence each element needs an additional type tag byte (though Message Pack encodes integers <128 with a single byte, so on average only ~50% of values get inflated this way).

Although my specific use-case is with Message Pack, I assume this issue would affect other formats with specific byte array encodings, though I don't know how common that is (as serde has first-class support in their data model, I'd guess it's not too uncommon?).

It's certainly possible to work around this downstream using `#[serde(with = "...")]` et-al, but I wanted to check whether this was considered when implementing #78, and considered desired behaviour, or if the bincode optimisation could be dropped for broader compatibility.

*Edit:* I guess this is basically a choice between known-size or known-type optimisations. A known-size byte array could benefit from both, but serde's data model doesn't support it.

**Comment by tarcieri** [comment]
Yes, this has come up several times, and is a very annoying deficiency of `serde` where there isn't a good way to describe an efficient byte serialization.

In hindsight, it would probably make sense to use something like [`serde_bytes`](https://crates.io/crates/serde_bytes).

Unfortunately this is a post-1.0 crate so this isn't something we can easily address without a breaking change.

**Comment by connec** [comment]
We are using `serde_bytes` for our workaround. That crate supplies its own `Deserialize` and `Serialize` traits, which it impls for [various byte array types](https://docs.rs/serde_bytes/0.11.5/serde_bytes/trait.Deserialize.html#implementors). The `#[serde(with = "serde_bytes")]` attribute will only work for types that implement the `serde_bytes::{Deserialize,Serialize}` traits, so we ended up with:

```rust
mod serialize {
    pub(super) fn deserialize<'de, D>(
        deserializer: D,
    ) -> Result<Signature, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let bytes: Vec<u8> = serde_bytes::deserialize(deserializer)?;
        let bytes_len = bytes.len();

        let bytes = bytes
            .try_into()
            .map_err(|_| serde::de::Error::invalid_length(bytes_len, &"a 64-byte array"))?;

        Ok(Signature::new(bytes))
    }

    pub(super) fn serialize<S>(
        signature: &Signature,
        serializer: S,
    ) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serde_bytes::serialize(signature.as_ref(), serializer)
    }
}
```

Would you be open to a PR that brings in a new feature flag, either to restore the bytes serialization directly or to include an impl of `serde_bytes::{Deserialize,Serialize}` to make it easier for consumer to opt-in? E.g. if `Signature` did impl `serde_bytes` traits then a workaround would be to simply wrap the `Signature`, e.g.

```rust
#[derive(serde::Deserialize, serde::Serialize)]
struct SignatureAsBytes {
    #[serde(with = "serde_bytes")]
    signature: Signature,
}
```

In fact – I decided to go ahead and see if this could be possible, so I've opened a PR for the 'add `serde_bytes`' option, in case it's worth consideration: #337.

**Comment by connec** [comment]
I also opened #338 with the approach of using a new feature to switch the implementation. I think this would be backwards compatible in general, unless `--all-features` is considered to be in scope for backwards compatibility.

I mostly wanted to see what the options would look like, so no harm done if introducing new features and the associated maintenance is undesired.

**Comment by connec** [comment]
Fixed by #337 🎉
